lib/adapters/GeoJSON.class.php

```diff
@@ -8,148 +9,173 @@
-      $polys[] = $this->arrayToPolygon($comp_array);
-    }
-    return new MultiPolygon($polys);
-  }
-
-  private function objToGeometryCollection($obj) {
-    $geoms = array();
-    if (empty($obj->geometries)) {
-      throw new Exception('Invalid GeoJSON: GeometryCollection with no component geometries');
-    }
-    foreach ($obj->geometries as $comp_object) {
-      $geoms[] = $this->objToGeom($comp_object);
-    }
-    return new GeometryCollection($geoms);
-  }
-
-  /**
-   * Serializes an object into a geojson string
-   *
-   *
-   * @param Geometry $obj The object to serialize
-   *
-   * @return string The GeoJSON string
-   */
-  public function write(Geometry $geometry, $return_array = FALSE) {
-    if ($return_array) {
-      return $this->getArray($geometry);
-    }
-    else {
-      return json_encode($this->getArray($geometry));
-    }
-  }
-
-  public function getArray($geometry) {
-    if ($geometry->getGeomType() == 'GeometryCollection') {
-      $component_array = array();
-      foreach ($geometry->components as $component) {
-        $component_array[] = array(
-          'type' => $component->geometryType(),
-          'coordinates' => $component->asArray(),
+    /**
+     * Given an object or a string, return a Geometry
+     *
+     * @param mixed $input The GeoJSON string or object
+     * @return object Geometry
+     * @throws Exception
+     */
+    public function read($input)
+    {
+
+        if (is_string($input)) {
+            $input = json_decode($input);
+        }
+        if (!is_object($input)) {
+            throw new Exception('Invalid JSON');
+        }
+        if (!is_string($input->type)) {
+            throw new Exception('Invalid JSON');
+        }
+
+        // Validate that type is valid
+        $validTypes = [
+            'Point',
+            'MultiPoint',
+            'LineString',
+            'MultiLineString',
+            'Polygon',
+            'MultiPolygon',
+            'GeometryCollection',
+            'Feature',
+            'FeatureCollection'
+        ];
+
+        if (!in_array($input->type, $validTypes)) {
+            throw new Exception('Invalid JSON');
+        }
+
+        // Check to see if it's a FeatureCollection
+        if ($input->type == 'FeatureCollection') {
+            $geoms = array();
+            foreach ($input->features as $feature) {
+                $geoms[] = $this->read($feature);
+            }
+            return geoPHP::geometryReduce($geoms);
+        }
+
+        // Check to see if it's a Feature
+        if ($input->type == 'Feature') {
+            return $this->read($input->geometry);
+        }
+
+        // It's a geometry - process it
+        return $this->objToGeom($input);
+    }
+
+    private function objToGeom($obj)
+    {
+        $type = $obj->type;
+
+        if ($type == 'GeometryCollection') {
+            return $this->objToGeometryCollection($obj);
+        }
+        $method = 'arrayTo' . $type;
```

Comment:
The validation you have above is great. But the actual issue happens here in this line. When `$type not in ['Point', 'LineString', 'Polygon', 'MultiPoint', 'MultiLineString', 'MultiPolygon']`... then we get a PHP Fatal error because the method (a)`$this->arrayToGeometryCollection()` or (b)`$this->arrayToPolyglot()` does not exist. The validation you have above will prevent (b) from happening, but (a) would still be possible.

Please validate here as well.